src/system.rs

```diff
@@ -31,3 +127,88 @@ mod unit_tests {
         assert_eq!(uptime_internal("12489513.08 22906637.29\n"), 12489513);
     }
 }
+
+/// Returns information about virtual memory usage
+///
+/// `/proc/meminfo` contains the virtual memory statistics
+pub fn virtual_memory() -> Option<VirtualMemory> {
+    let data = read_file(Path::new("/proc/meminfo")).unwrap();
+    let mem_info = make_map(&data);
+
+    let mut virtual_memory: VirtualMemory = Default::default();
+
+    virtual_memory.total = get_deref!(mem_info, "MemTotal:");
+    virtual_memory.free = get_deref!(mem_info, "MemFree:");
+    virtual_memory.buffers = get_deref!(mem_info, "Buffers:");
+    virtual_memory.cached = get_deref!(mem_info, "Cached:");
+    virtual_memory.active = get_deref!(mem_info, "Active:");
+    virtual_memory.inactive = get_deref!(mem_info, "Inactive:");
+
+    // MemAvailable was introduced in kernel 3.14. The original psutil computes it if it's not
+    // found, but since 3.14 has already reached EOL, let's assume that it's there.
+    virtual_memory.available = get_deref!(mem_info, "MemAvailable:");
+
+    // Shmem was introduced in 2.6.19
+    virtual_memory.shared = get_deref!(mem_info, "Shmem:");
+
+    virtual_memory.used = virtual_memory.total - virtual_memory.free - virtual_memory.cached -
+        virtual_memory.buffers;
+    virtual_memory.percent = (virtual_memory.used as f32 / virtual_memory.total as f32) * 100.0;
+
+    Some(virtual_memory)
+}
+
+/// Returns information about swap memory usage
+///
+/// `/proc/meminfo` and `/proc/vmstat` contains the information
+pub fn swap_memory() -> Option<SwapMemory> {
+    let data = read_file(Path::new("/proc/meminfo")).unwrap();
+    let swap_info = make_map(&data);
+
+    let vmstat = read_file(Path::new("/proc/vmstat")).unwrap();
+    let vmstat_info = make_map(&vmstat);
+
+    let mut swap_memory: SwapMemory = Default::default();
+
+    swap_memory.total = get_deref!(swap_info, "SwapTotal:");
+    swap_memory.free = get_deref!(swap_info, "SwapFree:");
+    swap_memory.sin = get_deref!(vmstat_info, "pswpin");
+    swap_memory.sout = get_deref!(vmstat_info, "pswpout");
+    swap_memory.used = swap_memory.total - swap_memory.free;
+    swap_memory.percent = (swap_memory.used as f32 / swap_memory.total as f32) * 100.0;
+
+    Some(swap_memory)
+}
+
+fn get_multiplier(fields: &mut Vec<&str>) -> Option<u64> {
+    if let Some(ext) = fields.pop() {
+        let multiplier = match ext {
+            "kB" => Some(1024),
+            _ => None,
+        };
+        fields.push(ext);
+
+        multiplier
+    } else {
+        None
+    }
+}
+
+fn make_map(data: &str) -> HashMap<&str, u64> {
```

Comment:
I struggle a bit to see what this is doing - it would be really helpful to have a couple of tests of this function explictly, so it's obvious how inputs map to outputs.